### PR TITLE
liskov substitution principle

### PR DESCRIPTION
**Pull request by gayansanjeewa**
* Derived classes must be substitutable for their base classes
* If B is a sub-class of A, then objects of class A may be replaced with
objects of class B.

### CODE REVIEW

File: 12-solid/3-liskov-substitution/0-notes.md [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+{
+    public function __invoke($foo)
+    {
+        // ...
+    }
+}
+
+class B extends A
+{
+    public function __invoke($foo)
```
Comment (Pragatheesh) [inline 12-solid/3-liskov-substitution/0-notes.md:23]: You can indicate its a overide, by phpdoc

File: 12-solid/3-liskov-substitution/0-notes.md [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    {
+        // ...
+    }
+}
+
+$x = new X();
```
Comment (Pragatheesh) [inline 12-solid/3-liskov-substitution/0-notes.md:29]: naming things ;P

Comment (Pragatheesh) [reply 12-solid/3-liskov-substitution/0-notes.md:29]: I would atleast say, ClassA, ClassB, ClassX

File: 12-solid/3-liskov-substitution/1-example/1-example.php [hunk 3]
```diff
@@ -0,0 +1,19 @@
+<?php
+
+class VideoPlayer
```
Comment (Pragatheesh) [inline 12-solid/3-liskov-substitution/1-example/1-example.php:3]: why is this a class and not an interface?

Comment (gayansanjeewa) [reply 12-solid/3-liskov-substitution/1-example/1-example.php:3]: We can screw up lsp in both ways, extension and implementation.
This example is about class extension, the other one is about implementation interfaces

File: 12-solid/3-liskov-substitution/0-notes.md (hunk 2 above)
Comment (gayansanjeewa) [reply 12-solid/3-liskov-substitution/0-notes.md:29]: it's an example just to demonstrate the above point
> If `B` is a sub-class of `A`, then objects of class `A` may be replaced with objects of class `B`.

File: 12-solid/3-liskov-substitution/0-notes.md (hunk 1 above)
Comment (gayansanjeewa) [reply 12-solid/3-liskov-substitution/0-notes.md:23]: I put it in the body itself.. want to keep it simple here

File: 12-solid/3-liskov-substitution/0-notes.md [hunk 4]
```diff
@@ -0,0 +1,32 @@
+# Liskov substitution principle
+
+![Illustration: Liskov substitution principle](../storage/3-intro-liskov-substitution-principle.png)
```
Comment (gayansanjeewa) [inline 12-solid/3-liskov-substitution/0-notes.md:3]: I kinda like this image more even I used the duck example

image

What do you guys think (say yes so my  inner nerd be satisfied :smiley: )

Comment (Pragatheesh) [reply 12-solid/3-liskov-substitution/0-notes.md:3]: For sure

Comment (gayansanjeewa) [reply 12-solid/3-liskov-substitution/0-notes.md:3]: we'll change it later if needed, for now I'll keep the existing one, i feel it's bit expressive than this one, let's see

**Pull request merged**